File: www/src/models/project.ts
export type Project = {
  id: string;
  title: string;
  slug: string;
  createdAt: string;
  updatedAt: string;
  description?: string;
  githubUrl?: string;
  wwwUrl?: string;
  content?: string;
  tags?: string[];
  featuredImageUrl?: string;
  videoUrl?: string;
};

export const reduceProject = (data: any): Project | null => {
  try {
    let blog: Project = {
      ...data
    };

    return blog;
  } catch (e) {
    console.debug('Error reducing project', e.message);
    return null;
  }
};

export const reduceProjects = (data: any[]): Project[] | null => {
  const blogs: Project[] = [];
  for (const item of data) {
    try {
      const blog = reduceProject(item);

      if (blog) blogs.push(blog);
    } catch (e) {
      console.debug('Error reducing project', e.message);
    }
  }
  return blogs;
};

File: api/schemas/project.py
from typing import List, Optional
from datetime import datetime

from schemas.base import BaseSchema, BaseModelSchema


class CreateProjectReq(BaseSchema):
    title: str
    description: str
    tags: List[str] = []
    content: str | None
    www_url: str | None = None
    github_url: str | None = None
    slug: str | None = None
    featured_image_url: str | None = None


# Inherit all fields from create project request schema
# plus add id, created_at, updated_at fields inherited
# from BaseSchema. Not used as request body schema for documentation
# purpose as to exclude id from examples in documentation
class ProjectSchema(CreateProjectReq, BaseModelSchema):
    pass


class CreateProjectRes(BaseSchema):
    status: str
    project: ProjectSchema | None


class DeleteProjectRes(BaseSchema):
    status: str
    delete_count: int


class UpdateProjectReq(BaseSchema):
    title: Optional[str] = None
    description: Optional[str] = None
    tags: Optional[List[str]] = None
    content: Optional[str] = None
    www_url: Optional[str] = None
    github_url: Optional[str] = None
    slug: Optional[str] = None
    updated_at: str | None = None
    created_at: str | None = None
    featured_image_url: str | None = None


class TagsRes(BaseSchema):
    tags: List[str] = []
